README.md
# nanoPPO

[![PyPI](https://img.shields.io/pypi/v/nanoPPO.svg)](https://pypi.org/project/nanoPPO/)
[![Changelog](https://img.shields.io/github/v/release/jamesliu/nanoPPO?include_prereleases&label=changelog)](https://github.com/jamesliu/nanoPPO/releases)
[![Tests](https://github.com/jamesliu/nanoPPO/workflows/Test/badge.svg)](https://github.com/jamesliu/nanoPPO/actions?query=workflow%3ATest)
[![Documentation Status](https://readthedocs.org/projects/nanoPPO/badge/?version=stable)](http://nanoPPO.readthedocs.org/en/stable/?badge=stable)
[![License](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/jamesliu/nanoPPO/blob/main/LICENSE)

nanoPPO is a Python package that provides a simple and efficient implementation of the Proximal Policy Optimization (PPO) algorithm for reinforcement learning. It is designed to support both continuous and discrete action spaces, making it suitable for a wide range of applications.

## Installation

You can install nanoPPO directly from PyPI using pip:

```bash
pip install nanoPPO
```

Alternatively, you can clone the repository and install from source:

```bash
git clone https://github.com/jamesliu/nanoPPO.git
cd nanoPPO
pip install .
```

## Usage

Here are examples of how to use nanoPPO to train an agent.

### On the MountaionCarContinuous-v0 environment:

```python
from nanoppo.train_ppo_agent import train_agent
import pickle

ppo, model_file, metrics_file = train_agent(
    env_name="MountainCarContinuous-v0",
    max_episodes=50,
    policy_lr=0.0005,
    value_lr=0.0005,
    vl_coef=0.5,
    checkpoint_dir="checkpoints",
    checkpoint_interval=10,
    log_interval=10,
    wandb_log=False,
)
ppo.load(model_file)
print("Loaded best weights from", model_file)
metrics = pickle.load(open(metrics_file, "rb"))
print("Loaded metrics from", metrics_file)
best_reward = metrics["best_reward"]
episode = metrics["episode"]
print("best_reward", best_reward, "episode", episode)
```

#### Use Custom LR Scheduler and Custom Policy

* Set Cosine Annealing Learning Rate Scheduler
* Set CausalAttention Policy instead of Linear Policy

```python
from nanoppo.train_ppo_agent import train_agent
from nanoppo.cosine_lr_scheduler import CosineLRScheduler
from nanoppo.policy.actor_critic_causal_attention import ActorCriticCausalAttention

lr_scheduler=CosineLRScheduler(
    learning_rate=config['cosine_lr'], 
    warmup_iters=config['cosine_warmup_iters'], 
    lr_decay_iters=config['cosine_decay_steps'], 
    min_lr=config['cosine_min_lr'])

policy_class = ActorCriticCausalAttention

ppo, model_file, metrics_file = train_agent(
    env_name=env_name,
    env_config = env_config, 
    max_episodes=config['max_episode'],
    stop_reward=config['stop_reward'],
    policy_class = policy_class,
    lr_scheduler=lr_scheduler,
    policy_lr=config['policy_lr'],
    value_lr=config['value_lr'],
    vl_coef=config['vl_coef'],
    betas = config['betas'],
    n_latent_var=config['n_latent_var'],
    gamma=config['gamma'],
    K_epochs=config['K_epochs'],
    eps_clip=config['eps_clip'],
    el_coef=config['el_coef'],
    checkpoint_dir=checkpoint_dir,
    checkpoint_interval=10,
    log_interval=10,
    wandb_log=wandb_log,
    debug=True)
```

### On the CartPole-v1 environment:

```python
from nanoppo.discrete_action_ppo import PPO
import gym

env = gym.make('CartPole-v1')
ppo = PPO(env.observation_space.shape[0], env.action_space.n)

# Training code here...
```
## Examples
See the [examples](https://github.com/jamesliu/nanoPPO/tree/main/examples) directory for more comprehensive usage examples.

examples/train_mountaincar.sh

```
python nanoppo/train_ppo_agent.py --env_name=MountainCarContinuous-v0 --policy_lr=0.0005 --value_lr=0.0005 --max_episodes=50 --vl_coef=0.5 --wandb_log
```
![mountaincar](https://github.com/jamesliu/nanoPPO/blob/main/assets/MountainCarContinuous-v0.png)

examples/train_pointmass1d.sh

```
python nanoppo/train_ppo_agent.py --env_name=PointMass1D-v0 --policy_lr=0.0005 --value_lr=0.0005 --max_episodes=50 --vl_coef=0.5 --wandb_log
```

examples/train_pointmass2d.sh

```
python nanoppo/train_ppo_agent.py --env_name=PointMass2D-v0 --policy_lr=0.0005 --value_lr=0.0005 --max_episodes=50 --vl_coef=0.5 --wandb_log
```

## Documentation

Full documentation is available [here](https://nanoppo.readthedocs.io/en/latest/).

## Contributing

We welcome contributions to nanoPPO! If you're interested in contributing, please see our [contribution guidelines](https://github.com/jamesliu/nanoPPO/blob/main/CONTRIBUTING.md).

## License

nanoPPO is licensed under the Apache License 2.0. See the [LICENSE](https://github.com/jamesliu/nanoPPO/blob/main/LICENSE) file for more details.

## Support

For support, questions, or feature requests, please open an issue on our [GitHub repository](https://github.com/jamesliu/nanoPPO/issues) or contact the maintainers.

## Changelog

See the [releases](https://github.com/jamesliu/nanoPPO/releases) page for a detailed changelog of each version.




docs/changelog.rst
.. _changelog:

===========
 Changelog
===========

.. _v0_15:

0.15 (2023-11-06)
-----------------

- Adding Actor Critic Causal Attention Policy
- Adding Cosine Annealing LR Scheduler
- Adding gradient and weight NAN and INF check
- Adding a debug flag when detecting NAN and INF 

.. _v0_14:

0.14 (2023-10-15)
-----------------

- Adding action rescaling: rescale actions
- Adding train_reward to metrics dictionary
- Adding version number in __init__.py
- Adding device configuration: facilitating both CPU and GPU configuration.
- Adding gradient clipping: introduced gradient clipping with clip grad norm to prevent updates that are too aggressive.

.. _v0_13:

0.13 (2023-09-19)
-----------------

- Creating a ppo agent training on continuous action space
- Adding two environments: `Pendulum-v1` and `MountainCarContinuous-v0`
- Creating two new environments: `PointMass1D-v0` and `PointMass2D-v0`
- Creating test cases: test_gae.py, test_returns_and_advantages.py, test_normalizer.py, test_random_utils.py , test_reward_scaler.py

.. _v0_11:

0.11 (2023-07-03)
-----------------

- Creating a ppo agent training on discrete action space
- Adding one environment: `CartPole-v1`

.. _v0_1:

0.1 (2023-07-01)
----------------

- Initial release to PyPI


train.py
import click
from nanoppo.ppo_agent import PPOAgent
from nanoppo.reward_shaper import TDRewardShaper, MountainCarAdvancedRewardShaper


@click.command()
@click.option("--project", default="ppoagent", help="Name of the project.")
@click.option(
    "--env_name",
    default="PointMass2D-v0",
    type=click.Choice(["PointMass1D-v0", "PointMass2D-v0", "Pendulum-v1", "MountainCarContinuous-v0"]),
    help="Name of the environment.",
)
@click.option("--epochs", default=1000, help="Number of training epochs.")
@click.option("--policy_lr", default=0.0005, help="Learning rate for policy network.")
@click.option("--value_lr", default=0.0005, help="Learning rate for value network.")
@click.option(
    "--rescaling_rewards", is_flag=True, default=False, help="Flag to rescale rewards."
)
@click.option("--scale_states", default="default", 
              type=click.Choice(["default", "env", "standard", "minmax", "robust", "quantile"]),
              help="Type of state scaling.")
@click.option("--max_timesteps", default=1000, help="Maximum number of timesteps for an episode.")
@click.option("--batch_size", default=200, help="Batch size for training updates.")
@click.option("--sgd_iters", default=4, help="Number of SGD iterations for training updates.")
@click.option("--use_gae/--no_use_gae", is_flag=True, default=True, help="Flag to use GAE.")
@click.option("--l1_loss", is_flag=True, default=False, help="Flag to use L1 loss for value/critic network.")
@click.option("--gamma", default=0.99, help="Discount factor.")
@click.option("--hidden_size", default=128, help="Hidden size for the neural network.")
@click.option(
    "--init_type",
    default="he",
    type=click.Choice(["he", "xavier"]),
    help="Initialization type for neural network weights.",
)
@click.option("--clip_param", default=0.2, help="Clipping parameter for PPO.")
@click.option("--vf_coef", default=0.5, help="Value function coefficient.")
@click.option("--entropy_coef", default=1e-2, help="Entropy coefficient.")
@click.option(
    "--max_grad_norm", default=1000, help="Maximum gradient norm for clipping."
)
@click.option(
    "--wandb_log", is_flag=True, default=True, help="Flag to log results to wandb."
)
@click.option(
    "--metrics_log", is_flag=True, default=False, help="Flag to log results to csv."
)
@click.option("--verbose", default=2, help="Verbosity level.")
@click.option("--checkpoint_interval", default=100, help="Checkpoint interval.")
@click.option("--checkpoint_dir", default="checkpoints", help="Path to checkpoint.")
@click.option("--log_interval", default=10, help="Logging interval.")
@click.option(
    "--resume_training", is_flag=True, default=False, help="Flag to resume training."
)
@click.option("--resume_epoch", default=0, type=int, help="Epoch to resume training from: <=0 for latest epoch.")
@click.option("--verbose", default=2, help="Verbosity level.")
# Add more configurations as needed


def cli(
    project,
    env_name,
    epochs,
    policy_lr,
    value_lr,
    rescaling_rewards,
    scale_states,
    max_timesteps,
    batch_size,
    sgd_iters,
    use_gae,
    l1_loss,
    gamma,
    hidden_size,
    init_type,
    clip_param,
    vf_coef,
    entropy_coef,
    max_grad_norm,
    wandb_log,
    metrics_log,
    checkpoint_interval,
    checkpoint_dir,
    log_interval,
    resume_training,
    resume_epoch,
    verbose,
):
    if use_gae:
        click.echo("Using GAE.")
    else:
        click.echo("Not using GAE.")
    # Configuration for the environment and the agent
    config = {
        "project": project,  # Name of the project
        "env_name": env_name,
        "env_config": None,
        "seed": None,
        "epochs": epochs,
        "rescaling_rewards": rescaling_rewards,
        "shape_reward": None,  # [None, SubClass of RewardReshaper]
        "scale_states": scale_states,  # [None, "env", "standard", "minmax", "robust", "quantile"]:
        "init_type": init_type,  # xavier, he
        "use_gae": use_gae,
        "tau": 0.95,
        "l1_loss": l1_loss,
        "sgd_iters": sgd_iters,
        "hidden_size": hidden_size,
        "max_timesteps": max_timesteps,
        "batch_size": batch_size,
        "gamma": gamma,
        "vf_coef": vf_coef,
        "clip_param": clip_param,
        "max_grad_norm": max_grad_norm,
        "entropy_coef": entropy_coef,
        "wandb_log": wandb_log,
        "metrics_log": metrics_log,
        "verbose": verbose,
        "checkpoint_interval": checkpoint_interval,
        "checkpoint_dir": checkpoint_dir,
        "log_interval": log_interval,
        "resume_training": resume_training,
        "resume_epoch": resume_epoch,
        "report_func": None,
    }
    optimizer_config = {
        "policy_lr": policy_lr,
        "value_lr": value_lr,
        "beta1": 0.9,
        "beta2": 0.999,
        "epsilon": 1e-8,
        "weight_decay": 0,
        "scheduler": None,
        "cosine_T_max": config["epochs"]
        * config["sgd_iters"],  # arbitrary value; you might want to adjust
        "exponential_gamma": 0.99,  # arbitrary value; you might want to adjust
    }

    agent = PPOAgent(config, optimizer_config)
    agent.train(epochs=epochs)


if __name__ == "__main__":
    cli()
